core/instance.go

```diff
@@ -18,53 +44,255 @@ type instanceTypeInformation struct {
+		//   original instance
+		// - volume size: each of the volumes should be at least as big as the
+		//   original instance's volumes
+
+		if attachedVolumesNumber > 0 {
+			logger.Println("Checking the new instance's ephemeral storage",
+				"configuration because the initial instance had attached",
+				"ephemeral instance store volumes")
+
+			if candidate.instanceStoreDeviceCount >= attachedVolumesNumber {
+				logger.Println("instance store volume count compatible,",
+					"continuing evaluation")
+			} else {
+				logger.Println("instance store volume count incompatible, skipping",
+					candidate.instanceType)
+				continue
+			}
+
+			if candidate.instanceStoreDeviceSize >= existing.instanceStoreDeviceSize {
+				logger.Println("instance store volume size compatible,",
+					"continuing evaluation")
+			} else {
+				logger.Println("instance store volume size incompatible, skipping",
+					candidate.instanceType)
+				continue
+			}
+
+			// Don't accept ephemeral spinning disks if the original instance has
+			// ephemeral SSDs, but accept spinning disks if we had those before.
+			if candidate.instanceStoreIsSSD ||
+				(candidate.instanceStoreIsSSD == existing.instanceStoreIsSSD) {
+				logger.Println("instance store type(SSD/spinning) compatible,",
+					"continuing evaluation")
+			} else {
+				logger.Println("instance store type(SSD/spinning) incompatible,",
+					"skipping", candidate.instanceType)
+				continue
+			}
+		}
+
+		if compatibleVirtualization(*i.VirtualizationType,
+			candidate.virtualizationTypes) {
+			logger.Println("virtualization compatible, continuing evaluation")
+		} else {
+			logger.Println("virtualization incompatible, skipping",
+				candidate.instanceType)
+			continue
+		}
+
+		// checking how many spot instances of this type we already have, so that
+		// we can see how risky it is to launch a new one.
+		spotInstanceCount := i.asg.alreadyRunningSpotInstanceCount(
+			candidate.instanceType, availabilityZone)
+
+		// We skip it in case we have more than 20% instances of this type already
+		// running
+		if spotInstanceCount == 0 ||
+			(*i.asg.DesiredCapacity/spotInstanceCount > 4) {
+			logger.Println(i.asg.name,
+				"no redundancy issues found for", candidate.instanceType,
+				"existing", spotInstanceCount,
+				"spot instances, adding for comparison",
+			)
+
+			filteredInstanceTypes = append(filteredInstanceTypes, candidate.instanceType)
+		} else {
+			logger.Println("\nInstances ", candidate, " and ", existing,
+				"are not compatible or resulting redundancy for the availability zone",
+				"would be dangerously low")
+
+		}
+
+	}
+	logger.Printf("\n Found following compatible instances: %#v\n",
+		filteredInstanceTypes)
+	return filteredInstanceTypes, nil
+
 }
 
-func (it *instance) isSpot() bool {
-	return (it.InstanceLifecycle != nil &&
-		*it.InstanceLifecycle == "spot")
+func (i *instance) getCheapestCompatibleSpotInstanceType() (*string, error) {
```

Comment:
Isn't it in this method that you are supposed to check that the SpotInstance that you get is in the same AZ as the ASG running? Ref: https://github.com/cristim/autospotting/issues/25